### PR TITLE
Add support for dollar count segment in filter

### PR DESCRIPTION
**Pull request by KenitoInc**
### Issues

*This pull request fixes issue #xxx.*

### Description

Add support for the following:

$filter=navProp/$count gt 10
$filter=navProp/$count($filter=prop gt 1) gt 2
$filter = collectionValuedProp/$count gt 10
$filter=collectionValuedProp/$count($filter=prop gt 1) gt 2

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs [hunk 1]
```diff
@@ -332,6 +332,39 @@ public virtual void ValidateConvertNode(ConvertNode convertNode, ODataValidation
             ValidateQueryNode(convertNode.Source, settings);
         }
 
+        /// <summary>
+        /// Override this method to restrict the '$count' inside the filter query.
+        /// </summary>
+        /// <remarks>
+        /// This method is intended to be called from method overrides in subclasses. This method also supports unit-testing scenarios and is not intended to be called from user code.
```
Comment (marabooy) [inline src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs:339]: Maybe this is not required?

**Review by marabooy — APPROVED** [review]

File: src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs [hunk 2]
```diff
@@ -224,8 +224,37 @@ private Expression BindCountNode(CountNode node)
...
+            }
+            else
+            {
+                whereMethod = ExpressionHelperMethods.EnumerableWhereGeneric.MakeGenericMethod(elementType);
+            }
+
+            // Bind the inner $filter clause within the $count segment.
+            // e.g Books?$filter=Authors/$count($filter=Id gt 1) gt 1
+            Expression filterExpression = null;
+            if (node.FilterClause != null)
```
Comment (xuzhg) [inline src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs:240]: It seems we don't need the else {} clause.
```C#
if (node.FilterClause != null)
{
    .....
   source = Expression.Call(null, ....);
}

countExpression = Expression.Call(null, countMethod, new[] { source });
```

**Review by xuzhg — APPROVED** [review]

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs (hunk 1 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs]: I have removed it.
#Resolved

File: src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs (hunk 2 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs:240]: #Resolved

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs [hunk 3]
```diff
@@ -332,6 +332,35 @@ public virtual void ValidateConvertNode(ConvertNode convertNode, ODataValidation
...
+        /// <summary>
+        /// Override this method to restrict the '$count' inside the filter query.
+        /// </summary>
+        /// <param name="countNode"></param>
+        /// <param name="settings"></param>
+        public virtual void ValidateCountNode(CountNode countNode, ODataValidationSettings settings)
+        {
+            if (countNode == null)
+            {
+                throw Error.ArgumentNull("convertNode");
```
Comment (gathogojr) [inline src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs:344]:
```suggestion
                throw Error.ArgumentNull("countNode");
```

File: src/Microsoft.AspNet.OData.Shared/Query/ParameterAliasNodeTranslator.cs [hunk 4]
```diff
@@ -398,7 +398,7 @@ public override QueryNode Visit(UnaryOperatorNode nodeIn)
         /// <returns>The translated node.</returns> 
         public override QueryNode Visit(CountNode nodeIn)
         {
-            return new CountNode((CollectionNode)nodeIn.Source.Accept(this));
+            return new CountNode((CollectionNode)nodeIn.Source.Accept(this), nodeIn.FilterClause, nodeIn.SearchClause);
```
Comment (gathogojr) [inline src/Microsoft.AspNet.OData.Shared/Query/ParameterAliasNodeTranslator.cs:401]: Did you cater for $search as well? If yes, maybe you should mention that in the PR description

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs [hunk 5]
```diff
@@ -332,6 +332,35 @@ public virtual void ValidateConvertNode(ConvertNode convertNode, ODataValidation
...
+            }
+
+            if (settings == null)
+            {
+                throw Error.ArgumentNull("settings");
+            }
```
Comment (gathogojr) [inline src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs:350]:
```suggestion
            }

```

**Review by gathogojr — APPROVED** [review]
Left a few comments, otherwise looks good

File: src/Microsoft.AspNet.OData.Shared/Query/ParameterAliasNodeTranslator.cs (hunk 4 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Query/ParameterAliasNodeTranslator.cs:401]: $search is not supported in WebAPI.
This code ensures we return the correct node after adding Parameter alias.

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs (hunk 5 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs:350]: #Resolved

File: src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs (hunk 3 above)
Comment (KenitoInc) [reply src/Microsoft.AspNet.OData.Shared/Query/Validators/FilterQueryValidator.cs:344]: #Resolved

**Pull request merged**